# Exercise 5: Airport Distances

In [5]:
import points
path = '../exercise04/airports.csv'
header, data = points.read_airport_csv(path)
header

['airport_id',
 'name',
 'city',
 'country',
 'iata_faa',
 'iaco',
 'latitude',
 'longitude',
 'altitude',
 'zone',
 'dst']

### 1. Function to get list of nearest neighbors using Eulidean distance.

See points/distance.py, function `get_neighbor_list`.  It takes two functions as arguments: the distance function (`euclidean_distance` or `great_circle_distance`) and the function for what type of neighbor to get (currect options are `nearest_airport` and `furthest_airport`; could be extended to more complex things like "closest foreign airport" or "airport with the most similar name").

In [6]:
points.get_neighbor_list?

In [7]:
# Get list of nearest neighbors for each airport (takes about a minute)
nn_list_euclidean = points.get_neighbor_list(header, data, neighbor_function=points.nearest_airport, dist_function=points.euclidean_distance)
print("Found nearest neighbors for all %d airports.  First entry:"%(len(nn_list_euclidean)) )
print(nn_list_euclidean[0])

Found nearest neighbors for all 6977 airports.  First entry:
(1, 5420, 0.42477695483866257)


### 2. Find the most isolated airport using Euclidean distance.

In [8]:
# Get most isolated airport from the list made in the previous cell
max_distance = -1
most_remote_airport_id_euclidean = -1
most_remote_airport_neighbor_id_euclidean = -1
for id1, id2, dist in nn_list_euclidean:
    if dist > max_distance:
        most_remote_airport_id_euclidean = id1
        most_remote_airport_neighbor_id_euclidean = id2
        max_distance = dist
#

print (points.get_airport_by_id(header, data, most_remote_airport_id_euclidean))
print ("Nearest neighbor (%.3f degrees away):"%(max_distance))
print (points.get_airport_by_id(header, data, most_remote_airport_neighbor_id_euclidean))

[2033, 'South Pole Station', "Stephen's Island", 'Antarctica', '', 'NZSP', -89.999997, 0.0, 9300.0, 127.0, 'U']
Nearest neighbor (51.010 degrees away):
[6436, 'Grytviken', 'Grytviken', 'South Georgia and the Islands', '', '', -54.16534, -36.30288, 0.0, -2.0, 'N']


### 3. Find the pair of airports that are separated by the largest Euclidean distance

Uses the function "furthest_airport" in points.distance, which is basically identical to the "nearest_airport" function used above.

In [9]:
fn_list_euclidean = points.get_neighbor_list(header, data, neighbor_function=points.furthest_airport, dist_function=points.euclidean_distance)
print("Found most distant airports from all %d airports.  First entry:"%(len(fn_list_euclidean)) )
print(fn_list_euclidean[0])

# Get most isolated airport from the list made in the previous cell
max_distance = -1
most_remote_airport_id_euclidean = -1
most_remote_airport_neighbor_id_euclidean = -1
for id1, id2, dist in fn_list_euclidean:
    if dist > max_distance:
        most_remote_airport_id_euclidean = id1
        most_remote_airport_neighbor_id_euclidean = id2
        max_distance = dist
#

apt1 = points.get_airport_by_id(header, data, most_remote_airport_id_euclidean)
apt2 = points.get_airport_by_id(header, data, most_remote_airport_neighbor_id_euclidean)
name_idx = header.index("name")
print ("Furthest pair of airports is %s and %s (%.3f degrees apart):"%(apt1[name_idx], apt2[name_idx], max_distance))
print (apt1)
print (apt2)

Found most distant airports from all 6977 airports.  First entry:
(1, 5959, 327.21204493103073)
Furthest pair of airports is Chatham Islands and Ugolny Airport (370.5 degrees apart):
[2010, 'Chatham Islands', 'Chatham Island', 'New Zealand', 'CHT', 'NZCI', -43.81, -176.457222, 43.0, 13.0, 'Z']
[6098, 'Ugolny Airport', 'Anadyr', 'Russia', 'DYR', 'UHMA', 64.73495, 177.741483, 194.0, 12.0, 'N']


### 4.1. Function to get list of nearest neighbors using Great Circle distance.

Same as above, but changing the distance function to haversine_distance (I could also leave the dist_function unspecified, as haversine is the default for all my distance-calculating functions).

In [10]:
# Get list of nearest neighbors for each airport (takes about a minute)
nn_list_great_circle = points.get_neighbor_list(header, data, neighbor_function=points.nearest_airport, dist_function=points.great_circle_distance)
print("Found nearest neighbors for all %d airports.  First entry:"%(len(nn_list_great_circle)) )
print(nn_list_great_circle[0])

Found nearest neighbors for all 6977 airports.  First entry:
(1, 5420, 0.4224520703921)


### 4.2. Find the most isolated airport using Great Circle distance.

In [11]:
# Get most isolated airport from the list made in the previous cell
max_distance = -1
most_remote_airport_id_great_circle = -1
most_remote_airport_neighbor_id_great_circle = -1
for id1, id2, dist in nn_list_great_circle:
    if dist > max_distance:
        most_remote_airport_id_great_circle = id1
        most_remote_airport_neighbor_id_great_circle = id2
        max_distance = dist
#

print (points.get_airport_by_id(header, data, most_remote_airport_id_great_circle))
print ("Nearest neighbor (%.3f degrees away):"%(max_distance))
print (points.get_airport_by_id(header, data, most_remote_airport_neighbor_id_great_circle))

[2657, 'Mataveri Intl', 'Easter Island', 'Chile', 'IPC', 'SCIP', -27.164792, -109.421831, 227.0, -6.0, 'S']
Nearest neighbor (23.376 degrees away):
[1979, 'Totegegie', 'Totegegie', 'French Polynesia', 'GMR', 'NTGJ', -23.079861, -134.890333, 7.0, -9.0, 'U']


### 4.3. Find the pair of airports that are separated by the largest Great Circle distance

In [13]:
fn_list_great_circle = points.get_neighbor_list(header, data, neighbor_function=points.furthest_airport, dist_function=points.great_circle_distance)
print("Found most distant airports from all %d airports.  First entry:"%(len(fn_list_great_circle)) )
print(fn_list_great_circle[0])

# Get most isolated airport from the list made in the previous cell
max_distance = -1
most_remote_airport_id_great_circle = -1
most_remote_airport_neighbor_id_great_circle = -1
for id1, id2, dist in fn_list_great_circle:
    if dist > max_distance:
        most_remote_airport_id_great_circle = id1
        most_remote_airport_neighbor_id_great_circle = id2
        max_distance = dist
#

apt1 = points.get_airport_by_id(header, data, most_remote_airport_id_great_circle)
apt2 = points.get_airport_by_id(header, data, most_remote_airport_neighbor_id_great_circle)
name_idx = header.index("name")
print ("Furthest pair of airports is %s and %s (%.3f degrees apart):"%(apt1[name_idx], apt2[name_idx], max_distance))
print (apt1)
print (apt2)

Found most distant airports from all 6977 airports.  First entry:
(1, 2556, 169.8267088592113)
Furthest pair of airports is Benito Salas and Sultan Mahmud Badaruddin Ii (179.948 degrees apart):
[2734, 'Benito Salas', 'Neiva', 'Colombia', 'NVA', 'SKNV', 2.95015, -75.294, 1464.0, -5.0, 'U']
[3289, 'Sultan Mahmud Badaruddin Ii', 'Palembang', 'Indonesia', 'PLM', 'WIPP', -2.89825, 104.699903, 49.0, 7.0, 'U']


### 5. Major Discrepancies

The most obvious difference is the pair of airports separated by the greatest distance- in Euclidean distance, which assumes the world is flat with four edges, the greatest distance is (roughly) from Siberia to New Zealand, with a distance of 370 degrees (clearly wrong, as no two points on a sphere can be more distant than 180 degrees).  The great circle answer is much nicer in that respect, with a distance of only 179.9 degrees. 

The most isolated airport also changes- it's less obvious that the Euclidean version is wrong, as it shows an antarctican airport, but there are multiple airports in antarctica.  The great circle answer of Easter Island is clearly better, though, as it's also (I believe) the most isolated chunk of land in the world.

### 6. Further investigations to try

- How many "farthest away airport" pairs are over 179 degrees?
- What's the least distant pair of farthest-away airports?
- What is the actual nearest airport (and distance) for South Pole Station (ID 2033)?
- What's the most distant airport from Chatham Islands (ID 2010)?
